### ISSUE TITLE
[Contributor Welcome] Implement C++ API version of torch.nn.TripletMarginLoss

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#27197]
## Context 

We would like to add the following two APIs to the C++ frontend:

*  `torch::nn::TripletMarginLoss` , which is the equivalent of Python API `torch.nn.TripletMarginLoss`.
* `torch::nn::functional::triplet_margin_loss`, which is the equivalent of Python API `torch.nn.functional.triplet_margin_loss`.

## Steps

* Add `torch::nn::TripletMarginLossOptions` to `torch/csrc/api/include/torch/nn/options/loss.h` (add this file if it doesn’t exist), which should include the following parameters (based on https://pytorch.org/docs/stable/nn.html#torch.nn.TripletMarginLoss)
    * `TORCH_ARG(double, margin) = 1.0`
    * `TORCH_ARG(double, p) = 2.0`
    * `TORCH_ARG(double, eps) = 1e-6`
    * `TORCH_ARG(bool, swap) = false`
    * `TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean`
    * NOTE: please make sure to add the same comments for the parameters as in the Python version. For example, for parameter `p` we should say `The norm degree for pairwise distance. Default: 2`.
* Add `torch::nn::functional::triplet_margin_loss(...)` in  `torch/csrc/api/include/torch/nn/functional/loss.h` (add this file if it doesn’t exist). The function should have the following signature:

```
namespace torch {
namespace nn {
namespace functional {

inline Tensor triplet_margin_loss(
    const Tensor& anchor,
    const Tensor& positive,
    const Tensor& negative,
    const TripletMarginLossOptions& options = {}) {
  ...
}

} // namespace functional
} // namespace nn
} // namespace torch
```

* Add `torch::nn::TripletMarginLoss` in `torch/csrc/api/include/torch/nn/modules/loss.h` (add this file if it doesn’t exist). The module’s constructor should take `torch::nn::TripletMarginLossOptions` as input (and store it in its internal options field), and the module’s forward method should have the following signature:

```
Tensor forward(
    const Tensor& anchor,
    const Tensor& positive,
    const Tensor& negative) {
  // Should call torch::nn::functional::triplet_margin_loss with all inputs and the module's options
}
```

* Add test for `torch::nn::TripletMarginLoss` in test/cpp/api/modules.cpp. We can just check that the output value for some simple cases are as expected.

* Add test for `torch::nn::functional::triplet_margin_loss(...)` in test/cpp/api/functional.cpp, using essentially the same tests as `torch::nn::TripletMarginLoss` .

## Helpful Resources

There are quite a few PRs for adding new functionals / new modules for the C++ API (the list of PRs is in https://github.com/pytorch/pytorch/issues/25883), which can serve as great references. Also please ping @yf225 on this issue if you encounter any problems.

## How do I claim this feature request?

Please comment in this issue if you are interested in working on it.

cc @yf225

**Comment by PyExtreme** [comment]
Hi @yf225  , I would like to contribute to this issue.

I am new to open source and want to contribute to PyTorch. This could be a great starting issue for me.

Thanks

**Comment by PyExtreme** [comment]
Hi @yf225 , I have added the two APIs to the C++ frontend. Can I please make a pull request so that you can review those changes?

**Comment by yf225** [comment]
@PyExtreme Thanks a lot and really appreciated the PR! I will review it today.

**Comment by PyExtreme** [comment]
Hi @yf225 , I have added tests as well.

Just a simple query.
There is no documentation for **_allclose_** function which can help in understanding it. Below is the link to it
https://pytorch.org/cppdocs/api/function_namespaceat_1a6718f09b724a8c39442d1a1585db19e1.html#exhale-function-namespaceat-1a6718f09b724a8c39442d1a1585db19e1

However, it is there in when I search for **_torch::allclose_**. Can I make a PR and add its doc?

**Comment by PyExtreme** [comment]
Hi @yf225 , I am getting an error while running the test. The link to the description is here:

https://discuss.pytorch.org/t/no-module-named-hypothesis-error-while-running-test/57603

It could be great if you could please help me with one.

**Comment by pbelevich** [comment]
@PyExtreme If you use conda, run `conda install hypothesis`,for pip I believe it should be `pip install hypothesis`

**Comment by chrahunt** [comment]
This was addressed in #27713.
